Simulate zero cross section for particles without processes

stognini opened the pull request:
This PR will transport a particle to the next geometry boundary until it escapes the world volume when no processes are assigned to it. This replicates Geant4's behavior when only the `G4Transportation` process is added to a given particle. Previously, the simulation would crash:
```
celeritas/track/SimTrackView.hh:
precondition failed: static_cast<bool>(sl.action) != (sl.step == numeric_limits<real_type>::infinity())
```

Since Celeritas has no equivalent `transportation` process, this is achieved by tweaking the pre- and along-step implementations:
- Pre-step sets action to `geo-boundary`, and sets both step limit and interaction MFP to `numeric_limits<real_type>::max()` for particles with no assigned processes.
- Along-step only calculates MFP if a process is assigned.

With this addition, #548 can probably be resumed?

stognini committed:
Fix typos

stognini committed:
Add transport bool to GeantPhysicsOptions

stognini committed:
Load transport option into PhysicsParamsOptions and PhysicsParamsScalars

stognini committed:
Update pre- and along-step interfaces to replicate G4Transportation

If a given particle has no processes assigned to it, and transport=true:
- Set step limit to infinity in pre-step
- Set mfp to infinty so that the particle reaches the next boundary

sethrj commented on src/celeritas/ext/GeantPhysicsOptionsIO.json.cc:
```diff
@@ -82,14 +82,17 @@ void from_json(nlohmann::json const& j, GeantPhysicsOptions& options)
     GPO_LOAD_OPTION(min_energy);
     GPO_LOAD_OPTION(max_energy);
     GPO_LOAD_OPTION(linear_loss_limit);
+    GPO_LOAD_OPTION(relaxation);
```
This is duplicated with `relaxation` above: likely some merge conflict resolution

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -142,6 +146,7 @@ operator==(GeantPhysicsOptions const& a, GeantPhysicsOptions const& b)
            && a.min_energy == b.min_energy
            && a.max_energy == b.max_energy
            && a.linear_loss_limit == b.linear_loss_limit
+           && a.relaxation == b.relaxation
```
```suggestion
```

sethrj commented on src/celeritas/ext/GeantImporter.cc:
```diff
@@ -408,14 +408,14 @@ import_materials(GeantImporter::DataSelection::Flags particle_flags)
-    ParticleFilter include_particle{process_flags};
-    ProcessFilter include_process{process_flags};
+    ParticleFilter include_particle{selection.processes};
+    ProcessFilter include_process{selection.processes};
```
I think you can revert the changes to this file?

sethrj commented on src/celeritas/global/alongstep/detail/AlongStepImpl.hh:
```diff
@@ -262,23 +262,34 @@ inline CELER_FUNCTION void apply_eloss(CoreTrackView const& track, EH&& eloss)
-            phys.interaction_mfp(mfp);
+            // Replicate G4Transportation
+            // Set infinite mfp so the particle reaches the next boundary
+            phys.interaction_mfp(numeric_limits<real_type>::infinity());
```
I don't think this should be in the along-step process. The infinite free path should be sampled during pre-step.

sethrj commented on src/celeritas/io/ImportModel.hh:
```diff
@@ -26,6 +26,7 @@ namespace celeritas
 enum class ImportModelClass
 {
     other,
+    unknown [[deprecated]] = other,  // remove in v0.4
```
Accidental addition from conflict resolution

sethrj commented on src/celeritas/phys/detail/PreStepActionImpl.hh:
```diff
@@ -62,10 +62,20 @@ inline CELER_FUNCTION void pre_step_track(celeritas::CoreTrackView const& track)
+    {
+        // Replicate G4Transportation
+        // Set step limit to infinity and exit
+        sim.reset_step_limit();
```
Instead of this, why not set to `{geo boundary, numeric_limits::max}`? I think then you wouldn't have to make any changes to the along-step kernel.

sethrj commented on src/celeritas/phys/detail/PreStepActionImpl.hh:
```diff
@@ -62,10 +62,20 @@ inline CELER_FUNCTION void pre_step_track(celeritas::CoreTrackView const& track)
+        // Set step limit to infinity and exit
+        sim.reset_step_limit();
+        sim.along_step_action()
+            = track.core_scalars().along_step_neutral_action;  // Needed?
```
Yeah it's needed.

sethrj reviewed (changes requested):
This also needs unit tests: it would have helped you a lot to add one before trying to validate your branch in your helper app. Update `celeritas/SimpleTestBase.cc` to use the null process and add a test to Stepper.test.cc.

stognini committed:
Address comments:

- Remove `transport` option; if no processes are added, simply transport
  to the next boundary
- Revert accidental additions from conflicts
- Simplify changes to pre- and alon-step implementations

stognini committed:
Add TestEm3 Compton to Stepper test

stognini committed:
Update test/CMakeLists

stognini committed:
Expand diagnostic test

stognini committed:
Update test; Use gtest FILTER in CMakeLists to call all Diagnostic tests

sethrj committed:
Revert incidental changes

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into null-process

sethrj committed:
Add no-Geant4 tests now that null process is available

sethrj reviewed (approved):
OK thanks for iterating with me on that! I think this is an easy "hack" for a good capability.

stognini commented:
Thanks for reviewing it. I think this way we replicate the behavior of G4 with as little code as possible---which seems cleaner than a fake process. It should work with any fake particle that has no assigned processes to it (like a celeritino) and we can add an option to kill and deposit energy locally for real particles in the future.

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into null-process
